docs/proposals/image-index.md

```diff
@@ -0,0 +1,150 @@
+  * Solving this problem goes beyond the scope of this proposal
+* This API is meant to only query the images/imagestreams not their usage in the Pods.
+* The classification is meant to be managed and used by cluster/registry administrators,
+  not by regular users
+
+## Proposed Workflow
+
+* As an admin I want to identify all *VENDOR* based images in the cluster:
+  * I create new `classification` using the *VENDOR* ImageStream(Tag)
+  * `ImageClassificationController` discovers all descentants of this Image (or any previous
+     version of this Image) in the cluster
+  * `ImageClassificationController` applies label/annotations to the Image(s) that include
+    reference to *VENDOR*
+
+* As an admin I want to group the classifications together, so I can send "key=value" queries, for example:
+  * I want to get all images with `os=rhel7` or `vendor=redhat` or `lang=python`:
+  * For that I create a new classification group named “os” and provide an “selector” that matches the labels
+    in the target “classification” objects.
+
+### Proposed API Changes
+
+For the implementation, two new API types are needed:
+
+```go
+// ImageClass provides a single Image "class" a context that a
+// certain group if Images shares. For example "rhel7" or "python".
+// To create an ImageClass, you have to point it to an initial
+// or "target" image first. From that point, the ImageClassController
+// will track the history of that Image.
+type ImageClass struct{
+		unversioned.TypeMeta
+		kapi.ObjectMeta
+
+		Current      ImageChain   
+		History      []ImageChain  // Should be in 'Status' ?
+}
+
+// ImageClassGroup groups the various ImageClass objects into a
+// logical group. For example a "vendor" group groups all ImageClass
+// objects that matches the defined selector.
+type ImageClassGroup struct {
+		unversioned.TypeMeta
+		kapi.ObjectMeta
+
+		Selector: map[string]string
+}
+
+// ImageChain represents an internal state of the Image layers and
+// the chains based on the image layers.
+type ImageChain struct {
+		Chains    []string
+}
+```
+
+The `ImageClass` object is needed to keep the historical informations about the
+image persistent as the in-memory image chain index will only contain the
+images and chains that are currently available in the cluster.
+
+In future this allows to see which images are based on old or insecure images
+and mark the affected images as outdated.
+
+
+### Image Chains Index
+
+The Image is represented by the list of its layers. To be able to identify the
+base image, we have to introspect the layers and calculate the "hash chains"
+for the image layers. We also need to keep this index in memory for fast
+access.
+
+For example:
+
+* The `base` image contains 3 layers: `AAA`, `BBB` and `CCC`
+  * For that we have 3 chains: `chain1(AAA)`, `chain2(AAA, BBB)`, `chain3(AAA, BBB, CCC)`
+
+* An `ancestor` image contains 4 layers: `AAA`, `BBB`, `CCC` and `DDD`
```

Comment:
> I'm not sure all ancestors are detected this way. It wouldn't handle when BBB is a metadata layer only, and thus (AAA, BBB, CCC) and (AAA, CCC) are actually identical.

I'm not sure if it is captured already by this document, but we've settled on ignoring metadata/empty layers. They are not present layer list in manifest schema 2 anyway.